### DISCUSSION TITLE
Set multiple in-memory keystores to handle different pModes

### DISCUSSION THREAD

**Discussion by binaradarsha** [phax/phase4#158] [General]
Hi,
We are going to handle multiple pModes in our as4 server project. i.e. BPC and Peppol. Looks we need two `AS4CryptoFactoryInMemoryKeyStore` objects due to it accepts only one key alias-password pair. 

In our custom `AS4Servlet` we have `AS4XServletHandler.setCryptoFactorySupplier(() -> cf)`. 
I am thinking if this `Supplier.get()` function would have a pMode parameter that injected for each document received like
`AS4XServletHandler.setCryptoFactorySupplier((pMode) ->  if(pMode == "BPC") cfBPC else cfPeppol)`.

Is my requirement is still achievable in a different way that I couldn't figured out yet, or needs to be implemented in phase4?

**Comment by phax** [comment]
So for sending it is pretty straighforward - you just place the right `IAS4CryptoFactory` in the right place.
However, when receiving it becomes a bit more tricky.

The "problem" I can see here is, that neither Peppol nor BPC use "PMode" in the intended way, due to the dynamic participant handling, Therefore I assume problem is similiar to #117 and the proposed solution in #121.

So if you implement you own `IAS4CryptoFactory` that also implements the interface `IAS4PModeAwareCryptoFactory` you can make the right choices, based on the context PMode.

However, while thinking about this - you need to make sure your solution is thread-safe, so that the conext PMode is e.g. stored in a ThreadLocal or so

**Reply by binaradarsha** [reply]
Hi, I wrote an implementation class for `IAS4PModeAwareCryptoFactory` with storing the PMode in a ThreadLocal variable.
It works fine when I send a Peppol document. But when I send a BPC document, PMode resolves to Peppol one. 
I debugged into bit deeper and found in `SOAPHeaderElementProcessorExtractEbms3Messaging.processHeaderElement()` that `Ebms3CollaborationInfo.agreementRef.pMode` is null eventhough there is a `bpc` related value. As a result default PMode is selected, probably the Peppol PMode as I see. 

Could this be a bug or because of we don't send some property from BPC client side? 
I suspect the `processSchemeId=null` in following code snippet. But as colleagues informed who has implemented our BPC client that property should be null for BPC.
```
Phase4BPCSender.bpcBuilder(Profile.BPC(), outboundId)
                .documentTypeID(docTypeId)
                .processID(processSchemeId, processId)
```

![AS4_Inbound-BPC_PMode_Issue1](https://github.com/phax/phase4/assets/19147983/7f21a0a5-38e0-44aa-ad84-345ad5f80101)

**Reply by phax** [reply]
If your `Phase4BPCSender` also (directly or indirectly) inherits from `AbstractAS4UserMessageBuilder` then it has a `.pmodeID(String)` method, with which you can provide an explicit PMode ID to use. That fields ends up in `CollaborationInfo/AgreementRef/@pmode` in the AS4 USerMessage

**Reply by binaradarsha** [reply]
As you said I set the `.pmodeID("BPC")` in sending side. So it appears in receiving side as below.

![AS4_Inbound-BPC_PMode_Issue2](https://github.com/phax/phase4/assets/19147983/6403337d-b8b3-45ee-a396-2556b35bf213)

But in a later step of `DefaultPModeResolver:79` it picks a default profile as Peppol.

![AS4_Inbound-BPC_PMode_Issue3](https://github.com/phax/phase4/assets/19147983/6719e695-de07-4f14-8439-dc4080e58c20)

What could be the reason?

**Reply by phax** [reply]
Hmmmm, the first few lines of the `DefaultPModeResolver` look like this:
```java
   IPMode ret = null;
    if (StringHelper.hasText (sPModeID))
    {
      // An ID is present - try to resolve this ID
      ret = aPModeMgr.getPModeOfID (sPModeID);
      if (ret != null)
        return ret;
    }
```

So it means, your PModeManager does not know the ID `bpc`. Have you created a "persistent" PMode? And what `IPModeManager` are you using by default? If you haven't changed anything, you are using the `PModeManagerInMemory` which only maintains the PModes in memory. For BPC however you need that PMode all the time. The easiest solution, is to always "create" the PMode on application startup. Alternatively you switch over to a `PModeManager` implementation that persists data on disk. hth

**Reply by binaradarsha** [reply]
Hi, as you said `PModeManagerInMemory` is used. When I further debugged into `aPModeMgr.getPModeOfID (sPModeID)` found that PMode is picked from a map called `m_aMap`. It has two entries with key as the operatorId issued from BPC/Peppol networks in format <operatorId>-<operatorId>, ie. `EIN::....-EIN::.....`, `PNO....-PNO....`. Value is a PMode object, and in both entries `PMode.m_sAgreement = urn:fdc:peppol.eu:2017:agreements:tia:ap_provider`. 

As I checked we haven't created a IPModeManager instance explicitly. Anyhow it's created and I am curious how two entries are populated? Is it from our keystore?

However my main question is how could BPC entry also got this Peppol agreement?

**Reply by phax** [reply]
Well okay, the `PModeManagerInMemory` is the default instance that is created via the `AS4ManagerFactoryInMemory` which is instantiated via `MetaAS4Manager`.
Honestly it is hard to make definitive assumptions about your setup, without looking at the code ;-) The dynamic PMode IDs are created via the default implementation of `IPModeIDProvider`.
The default BPC `IPModeProvider` implementation stems from the `AS4BPCProfileRegistarSPI`, assuming that BPC also uses the "dynamic PMode concept" as OpenPeppol does. Why do you want a single PMode with a single combination of sender and receiver for BPC by the way? Are you including the `phase4-profile-bpc` artifact in your code?

In general I would recommend to setup separate instances for separate networks. The detailed configuration might become quite messy, as can be seen here....

**Reply by binaradarsha** [reply]
I will look into with PModeProvider and related stuff you have mentioned. Yes, we are using `phase4-profile-bpc` artifact in our project.
Our organization needs to handle all AS4 profiles by a single service. We have to keep up with organizational decisions and so need to solve this problem somehow.

**Reply by phax** [reply]
Okay, eventually you can provide multiple "endpoints" (aka Servlets) to deal with Peppol AS4 messages on "/as4peppol" and receive BPC stuff at "/as4bpc"? That way you can easily configure 2 different receiving handlers with different setups etc.

**Reply by binaradarsha** [reply]
Yeah, that is also a good alternative. Once initially I thought to have separate Servlets, but then considered having two urls would be problematic. But looks we just need to register "/as4bpc" in BPC network and "/as4peppol" in Peppol network. I will discuss with our colleagues who have more understanding regarding this stuff.
Thank you!

**Reply by binaradarsha** [reply]
Hi, we moved forward with a single url "/as4" using the workaround below in our AS4Servlet.

```
final AS4XServletHandler servletHandler = new AS4XServletHandler();

servletHandler.setPModeResolver((sPModeID, sService, sAction, sInitiatorID, sResponderID, sAgreementRef, sAddress) -> {
    if (sPModeID.equals("BPC")) {
        return BPCPMode.createBPCPMode(sInitiatorID, sResponderID, sAddress, (i, j) -> sPModeID, false);
    } else if (sPModeID.equals("PEPPOL")) {
        return PeppolPMode.createPeppolPMode(sInitiatorID, sResponderID, sAddress, (i, j) -> sPModeID, false);
    }
    return null;
});
```

sPModeID is one we set in our Phase4BPCSender/Phase4PeppolSender.pmodeID() method in client side.

In our IAS4PModeAwareCryptoFactory implementation we used a ThreadLocal variable to store the IPMode object as you have suggested. When a thread pool is used, for handling a message a thread is picked from the pool, it will have its own copy of IPMode object, finally when finished message handling it will be released to the pool with keeping the IPMode object. Then if that same thread is picked again for another message handling, it's IPMode object is overridden and we don't need to manually remove it beforehand the thread is released to the pool. 
Is this correct or do you recommend to manually remove the IPMode object?
